### PR TITLE
Add missing type hints to synapse.http.

### PR DESCRIPTION
**Pull request by clokep**
This starts adding some missing type hints to `synapse.http`. There's still quite a bit to do but wanted to continue making baby steps!

I wasn't sure if I should add these to `mypy.ini` or not, but listing individual files seems painful.

### CODE REVIEW

File: synapse/http/server.py [hunk 1]
```diff
@@ -271,7 +276,7 @@ async def _async_render_wrapper(self, request: SynapseRequest):
             f = failure.Failure()
             self._send_error_response(f, request)
 
-    async def _async_render(self, request: Request):
+    async def _async_render(self, request: SynapseRequest) -> Optional[Tuple[int, Any]]:  # type: ignore[return]
```
Comment (squahtx) [inline synapse/http/server.py:279]: What complaint is being suppressed here?

Comment (clokep) [reply synapse/http/server.py:279]: Seems like it isn't needed. Not sure why I added it!

**Review by squahtx — APPROVED** [review]
lgtm!

**Comment by DMRobertson** [comment]
I think a lot of this is duplicated within Sydent and Sygnal---would be keen to see stuff migrated to matrix-python-common if you see a good opportunity to do so!

**Pull request merged**